kratos/python_scripts/analysis_stage.py

```diff
@@ -211,6 +211,20 @@ def ChangeMaterialProperties(self):
         """this function is where the user could change material parameters as a part of the solution step """
         pass
 
+    def Save(self, serializer : KratosMultiphysics.StreamSerializer):
+        """Serializes current analysis stage instance
+
+        This method is intended to make the class pure Python (pickable). This means serialize all the Kratos objects with the given serializer to then destroy them.

```

Comment:
If you don't destroy (i.e. assigning them to `None`) the non pure Python object the class wouldn't be pickable.

Reply:
I don't get the point. A typical use case would be save and continue. No?

Reply:
Yes. But we are using Python `pickle` for doing so. Pickle is only capable to save / load pure Python types. Hence, we need to convert all Kratos objects to a pure Python type (this is what we do with our stream serializer). Then you "destroy" these objects in order to make the class Pickable. Let me try to write an example.

We have a `MyStage` analysis stage class with whatever Kratos object e.g. a Newton-Raphson instance. At this point, the class is not pickable as `self.strategy` cannot be dump. In order to do so, we call the stream serializer within the `Save` method of `MyClass` to create a string "representing" it. At this point, the `MyStage` is still not pickable as `self.strategy` is not a Python type but a Kratos one. Then, we need to do `self.strategy = None` after the stream serialization. At this point the clas IS pickable. This is what I meant by destroy. In pseudocode will be something like
```py
def Save(self, serializer: KratosMultiphysics.StreamSerializer):
    serializer.Save(self.strategy) # class is not pickable yet
    self.strategy = None # class is now pickable
```

Reply:
> Pickle is only capable to save / load pure Python types. 

https://pybind11.readthedocs.io/en/latest/advanced/classes.html#pickling-support

Reply:
But the documentation you are referring don't says anything about making it null. It only says that `__setstate__` and `__getstate__` methods to be defined. Would you please give me the reference about making it null?

Reply:
Here's my take on it:

An object is `pickle`able if it defines `__setstate__` and `__getstate__`, **or** all its member variables are `pickle`able. Note the recursion here.

## Before `AnalysisStage::Save`

The analysis stage may have objects coming from pybind, such as solvers, processes, etc. By default, `pickle` cannot handle these because classes bound from C++ are only picklable if someone explicitly added support for it (i.e.: defined `__setstate__` and `__getstate__` in the bindings).

## In `AnalysisStage::Save`, before `pickle`

Before passing the analysis stage on to `pickle`, these members coming from C++ need to be dealt with. I'm guessing @rubenzorrilla proposes to use the kratos serializers to take over `pickle`'s job for these classes. After the kratos serializers did their job, the C++ classes are captured by the serializer so they can be removed from the analysis stage to make it `pickle`able (`self.cpp_class_instance = None`).

## `pickle`

The analysis stage is now free of un`pickle`able members, so it can be safely passed on to `pickle`.

##

My problem with this approach is that the analysis stage is left in an invalid state after `AnalysisStage::Save`. This means that at every checkpoint, we must not only `Save` the analysis stage but also `Load` it back in => i.e.: perform a restart every time.

Reply:
Just to say that I added more details to the docstring.

Reply:
> Here's my take on it:
> 
> An object is `pickle`able if it defines `__setstate__` and `__getstate__`, **or** all its member variables are `pickle`able. Note the recursion here.
> 

This is what I understood

> ## Before `AnalysisStage::Save`
> The analysis stage may have objects coming from pybind, such as solvers, processes, etc. By default, `pickle` cannot handle these because classes bound from C++ are only picklable if someone explicitly added support for it (i.e.: defined `__setstate__` and `__getstate__` in the bindings).
> 
> ## In `AnalysisStage::Save`, before `pickle`
> Before passing the analysis stage on to `pickle`, these members coming from C++ need to be dealt with. I'm guessing @rubenzorrilla proposes to use the kratos serializers to take over `pickle`'s job for these classes. After the kratos serializers did their job, the C++ classes are captured by the serializer so they can be removed from the analysis stage to make it `pickle`able (`self.cpp_class_instance = None`).

This is exactly my question: Why we should do self.cpp_class_instance=None? Would you please point me to the documentation in this respect?

> 
> ## `pickle`
> The analysis stage is now free of un`pickle`able members, so it can be safely passed on to `pickle`.
> 
> My problem with this approach is that the analysis stage is left in an invalid state after `AnalysisStage::Save`. This means that at every checkpoint, we must not only `Save` the analysis stage but also `Load` it back in => i.e.: perform a restart every time.

And this is my direct concern. In the case of this PR the stage is already deleted. But this limits the usage very much.

Reply:
> This is exactly my question: Why we should do `self.cpp_class_instance=None`? Would you please point me to the documentation in this respect?

I can't point you to the docs, because this is just one possible roundabout solution to a problem we created for ourselves. The followings are just as valid solutions:
`self.cpp_class_instance = 0`
`self.cpp_class_instance = "whatever"`
`self.cpp_class_instance = 6.0 * 1e24`
`self.cpp_class_instance = anything_that_is_pickleable`
`del self.cpp_class_instance`

The point is to remove any un`pickle`able member.

The proper (albeit more laborious) solution is to make our C++ types `pickle`able.

Reply:
the thing is more delicate than simple copy-pasting.

Imagine you want to store a Model and then some other python object.

one way of doing this could be to create a serializer object in the pickle pybind call, write it to a string and then write the resulting char array to pickle.

the problem of this approach is that if we now want to have a different save for the second object, we don't have access to the serializer used to save the first object. We could generate a second StreamSerializer insance, however this implies that common things (like the kernel) would be serialized twice (and differently) in the first object and in the second.

What i am trying to say is that we need to first serialize all of the cpp objects USING THE FIRST SERIALIZER, and then write the serialized object using pickle. That is exactly what we are doing.
no doubts we could have a global serializer used in python ... however this complicates things quite a lot.

having said this, this is an implementation detail. If you like, let's open a different thread to discuss about that